rfcs/text/0000-matrix-service.md

```diff
@@ -0,0 +1,406 @@
+
+### User Stories
+
+As a data consumer, I would like to access gene expression data from the DCP so that I can perform downstream integrative analysis across HCA experiments.
+
+- I would like to be able to receive an expression matrix comprised of cells that satisfy a set of metadata conditions.
+- I would like to receive metadata providing the experimental and scientific context about the cells in my matrix.
+- I would like to receive my matrix in a format that is compatible with standard single-cell analysis tools.
+
+## Scientific "guardrails"
+
+*Describe recommended or mandated review from HCA Science governance to ensure that the RFC addresses the needs of the scientific community.*
+
+## Detailed Design
+
+### Overview
+
+The Matrix Service enables on-the-fly expression matrix generation via a [RESTful API](https://matrix.data.humancellatlas.org/). Through user studies and iterative development over the past year, four major requirements of matrices have been identified and must be met. One, matrices must enable integrative analysis across data in the HCA and the Matrix Service determines that this integration should be at the granularity of individual cells. Second, matrices should include rich metadata that users are interested in so that they can better understand the data they're working with. Third, matrices should be compatible with downstream single cell analysis tools, for example in `Python` and `R`, to further enable downstream analysis of HCA data by the wider community. Finally, performance of serving matrices should be prioritized as the scale of HCA data is anticipated to reach or exceed the order of 10<sup>7</sup> cells. The Matrix Service API is designed to fulfill these requirements in order to provide access to HCA data in a format that enables and encourages downstream integrative analysis.
+
+The default and recommended file format for matrices generated by the service is [loom](http://loompy.org/), an HDF5 file format that is optimized for storing large omics datasets such as expression matrices and designed for consumability, for example, by exporting metadata with matrices. The Matrix Service achieves its first three goals by delivering matrices in `loom`, and other supporting formats, via an expressive API that enables users the ability to specify individual cells and desired metadata to generate a matrix for. For a detailed example, see [this user vignette](https://nbviewer.jupyter.org/github/HumanCellAtlas/matrix-service/blob/master/docs/HCA%20Matrix%20Service%20to%20Scanpy.ipynb#HCA-Matrix-Service-Loom-Output) describing a `loom` matrix generated by the service. The fourth requirement is satisfied by a lazy cache of matrices providing constant time delivery of previously requested matrices.
```

Comment:
How fast is "constant time"? I think this is a very computer science specific term so maybe add some concreteness to it?

Reply:
I see three locations where "constant time" is mentioned so maybe correct throughout?

Reply:
To answer your question, minutes if not less. Good call out, will update

Reply:
Done